// expire/lib/expire.js

var units = {
    year:    31536000
  , month:   2592000
  , week:    604800
  , day:     86400
  , hour:    3600
  , minute:  60
  , second:  1
}

var values = [
    'zero'
  , 'one'
  , 'two'
  , 'three'
  , 'four'
  , 'five'
  , 'six'
  , 'seven'
  , 'eight'
  , 'nine'
  , 'ten'
  , 'eleven'
  , 'twelve'
  , 'thirteen'
  , 'fourteen'
  , 'fifteen'
  , 'sixteen'
  , 'seventeen'
  , 'eighteen'
  , 'nineteen'
  , 'twenty'
]

values = values.map(function(value, index) {
    return [value, ''+index]
})

function getValue(v) {
    var res = 0
    for (var i=0, len=values.length; i<len; i++) {
        if (values[i].indexOf(v) !== -1) {
            return i
        }
    }
    return res
}

function computeSeconds(value, unit) {
  unit = unit.replace(/s$/i, '')
  unit = parseInt(units[unit.toLowerCase()])
  value = parseInt(getValue(value.toLowerCase()))
  if (!value || !unit) return 0
  return value * unit
}

function getSeconds(str) {

  if (!str || typeof str !== 'string') {
    return 0
  }

  var subs = str.split(/\sand\s/)
  
  subs = subs.map(function(i) { 
    return i.split(' ')
  })

  var sum = 0

  subs.forEach(function(sub) {
    sum += computeSeconds(sub[0], sub[1])
  })

  return sum
}

function setExpiration(res, time) {

    if (!time) {
        time = res
        res = {}
    }

    var seconds = typeof time === 'number' ? time : getSeconds(time)
    var date = new Date()
    date.setHours(date.getHours() + (seconds/3600))

    res['Expires'] = date.toUTCString()
    res['Cache-Control'] = 'max-age='+seconds

    return res

}

module.exports = {
    units:units,
    values:values,
    getSeconds:getSeconds,
    setExpiration:setExpiration
}
